### PR TITLE
Add script to generate hard filtered samples

### PR DESCRIPTION
**Pull request by mike-w-wilson**
This script adds hard filtering. I left what I considered to be TODOs commented in the code, e.g. determine threshold cut offs through visual inspection. I've tested this code on the test VDS so all samples are failing given the small number of variants.

This also adds a failed fingerprinting table resource.

### CODE REVIEW

**Comment by jkgoodrich** [comment]
There are still some TODOs on this PR because I'm not sure what sample QC hard filters we will actually use. I think it's overkill to add min and max options for all possible sample QC metrics, and I would like to see the full gnomAD v4 sample QC metric distributions. This file will need to change after we determine the best hard filtering for gnomAD v4

**Comment by jkgoodrich** [comment]
@klaricch will first look over just the sample QQ generation

**Comment by klaricch** [comment]
> One question is, should the sex imputation addition just be its own argument that makes the final file, with an intermediate file made without it rather than an option that determines what is in the final file?

I think I like the former. How would that option for the latter be structured?

**Review by klaricch — CHANGES_REQUESTED** [review]

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    interval_coverage,
+    sex,
+)
+
+logging.basicConfig(format="%(levelname)s (%(name)s %(lineno)s): %(message)s")
+logger = logging.getLogger("hard_filters")
+logger.setLevel(logging.INFO)
+
+
+def compute_sample_qc(n_partitions: int = 1000, test: bool = False) -> hl.Table:
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:33]: arbitrary. I was going for a round number and trying to make it more partitions given the change in sample size from v3. It was 100 for the genomes in v3 which was ~150,000 samples (prior to releasable filtering) so maybe 500 would be better?

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    calling_interval_padding = args.calling_interval_padding
+
+    try:
+        if args.sample_qc:
+            compute_sample_qc(
+                n_partitions=args.sample_qc_n_partitions, test=args.test
+            ).write(get_sample_qc(test=args.test).path, overwrite=args.overwrite)
+
+        if args.compute_coverage:
+            if args.test:
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:217]: no, but I think you are right, it using multiple times, let's just define at the top and use the variable

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    finally:
+        logger.info("Copying log to logging bucket...")
+        hl.copy_log(get_logging_path("hard_filters"))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--overwrite",
+        help="Overwrite all data from this subset (default: False)",
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:283]: Good catch, just a copy paste error

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+        help="Computes samples to be hard-filtered. NOTE: Cutoffs should be determined by visual inspection of the metrics.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--include-sex-filter",
+        help="If sex filters should be included in hard filtering.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--include-low-cov-filter",
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:330]: No, I ended up removing it, I didn't think it was needed. I will remove the flag too. Thanks!

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+        "--max-n-snp",
+        type=float,
+        default=3.75e6,
+        help="Filtering threshold to use for the maximum number of SNPs. Default is 3.75e6.",
+    )
+    hard_filter_args.add_argument(
+        "--min-n-snp",
+        type=float,
+        default=2.4e6,
+        help="Filtering threshold to use for the minimum number of SNPs. Default is 2.4e6.",
```
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:354]: it's the minimum cutoff though for `n_snp`, the maximum cutoff argument is above

**Comment by jkgoodrich** [comment]
> > One question is, should the sex imputation addition just be its own argument that makes the final file, with an intermediate file made without it rather than an option that determines what is in the final file?
> 
> I think I like the former. How would that option for the latter be structured?

As is it will always write using

```
    .write(
                get_checkpoint_path(
                    f"test_gnomad.exomes.hard_filtered_samples.ht"
                )
                if test
                else hard_filtered_samples.path,
                overwrite=args.overwrite,
            )
```

No matter what the value of `args.include_sex_filter` is. But I'm wondering if instead we should include that in the name of the file if it's not set?

File: gnomad_qc/v4/sample_qc/hard_filters.py [hunk 6]
```diff
@@ -264,12 +268,19 @@ def main(args):
...
-                args.test,
-            ).write(hard_filtered_samples.path, overwrite=args.overwrite)
+                args.min_n_over_gq,
+                args.min_gq,
+                args.min_n_over_dp,
+                args.min_dp,
+                test,
+            ).write(
+                get_checkpoint_path(
+                    f"test_gnomad.exomes.hard_filtered_samples.ht"
```
Comment (klaricch) [inline gnomad_qc/v4/sample_qc/hard_filters.py:278]:
```suggestion
                    "test_gnomad.exomes.hard_filtered_samples"
```

File: gnomad_qc/v4/sample_qc/hard_filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (klaricch) [reply gnomad_qc/v4/sample_qc/hard_filters.py:360]: oops, meant to edit the arg above

File: gnomad_qc/v4/sample_qc/hard_filters.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (klaricch) [reply gnomad_qc/v4/sample_qc/hard_filters.py]: 500 sounds good

**Review by klaricch — CHANGES_REQUESTED** [review]

File: gnomad_qc/v4/sample_qc/hard_filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad_qc/v4/sample_qc/hard_filters.py:360]: actually, I will just change all of those below to say maximum

**Review by klaricch — APPROVED** [review]
lgtm!

**Pull request merged**